src/Controller/ControllerFactory.php

```diff
@@ -38,23 +38,23 @@ class ControllerFactory implements ControllerFactoryInterface
      * @param \Psr\Http\Message\ServerRequestInterface $request The request to build a controller for.
      * @return \Cake\Controller\Controller
      * @throws \Cake\Http\Exception\MissingControllerException
-     * @throws \ReflectionException
      */
     public function create(ServerRequestInterface $request): Controller
     {
         $className = $this->getControllerClass($request);
-        if (!$className) {
+        if ($className === null) {
             $this->missingController($request);
         }
-        /** @var string $className */
-        $reflection = new ReflectionClass($className);
-        if ($reflection->isAbstract() || $reflection->isInterface()) {
+
+        $controller = null;
+        try {
+            /** @psalm-suppress UndefinedClass */
+            $controller = new $className($request);
+        } catch (Error $e) {
             $this->missingController($request);
```

Comment:
Before, you had a 2nd argument to rethrow $e into the chain. Might be worth resurrecting that change from before.

Reply:
Can't you just pretend you didn't see that commit? :stuck_out_tongue:

Reply:
`Error` can be thrown not only because controller is not instantiable. Many things can happen in many parts of constructor.

I would stay with reflection. Only maybe simplify to `ReflectionClass::isInstantiable()`.

Reply:
@garas One could argue that once the controller is instantiable, that anything else that goes wrong warrants an internal error 500, since then there is a real issue.
But `ReflectionClass::isInstantiable()` sounds also like a good idea.